chore: add upgrade guide for v.1.5 (#507)
cartesi/rollups-node: docs/upgrade_to_1_5.md

> [!IMPORTANT]
> It's highly recommended that application owners upgrade their instances of the Cartesi Rollups Node to `v1.5.0`, had their applications being affected by the issue above or not.

## What has changed

The way the Cartesi Rollups Node calculates epochs has changed starting on `v1.5.0`.
This procedure is now based on the `CARTESI_EPOCH_LENGTH` environment variable instead of `CARTESI_EPOCH_DURATION`.

The value of `CARTESI_EPOCH_LENGTH` (blocks) may be derived from `CARTESI_EPOCH_DURATION` (seconds) as follows:

`CARTESI_EPOCH_LENGTH = CARTESI_EPOCH_DURATION/BLOCK_TIME`

Where `BLOCK_TIME` corresponds to the duration it takes to generate a block in the target network.

> [!TIP]
> Suppose a block is created every `12` seconds and `CARTESI_EPOCH_DURATION` is set to `86400` seconds (24 hours).
>
> So, `CARTESI_EPOCH_LENGTH = 86400/12 = 7200`

Check the [`CHANGELOG`](./CHANGELOG.md) and the [configuration](./docs/config.md) for more details.

## How to upgrade

> Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v.1.5.0`.

Reviewer: ```suggestion
Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v1.5.0`.
```

This is the simplest way to perform the upgrade.

In order to do so, just update the application configuration considering [what has changed](#what-has-changed) and upgrade the version of the Cartesi Rollups Node as usual.

> [!CAUTION]
> A redeployment will create a new instance of the application from scratch.
> All previous inputs, outputs and claims will remain associated to the previous application address, funds included.

Alternatively, if the application owner wants to keep the application inputs, they may choose to replace the _History_ being used by the application's _Authority_ instead.
This process is a little bit more involved and is described in the next section.

## Steps to replace an Application's _History_

> [!CAUTION]

> > Instances of the [History](https://github.com/cartesi/rollups-contracts/blob/v1.4.0/onchain/rollups/contracts/history/History.sol) contract from [rollups-contracts v.1.4.0](https://github.com/cartesi/rollups-contracts/releases/tag/v1.4.0) may be used simultaneously by several applications through their associated _Authority_ instance.

Reviewer: ```suggestion
> Instances of the [History](https://github.com/cartesi/rollups-contracts/blob/v1.4.0/onchain/rollups/contracts/history/History.sol) contract from [rollups-contracts v1.4.0](https://github.com/cartesi/rollups-contracts/releases/tag/v1.4.0) may be used simultaneously by several applications through their associated _Authority_ instance.
```

> Application owners must consider that and exercise care when performing the steps listed below.

To keep the application inputs, before performing the upgrade to version `v1.5.0`, proceed as described in the sub-sections below.

### 1. Instantiate a new _History_

This is a two-step process.
First, the address of the new _History_ instance should be calculated.

> [!NOTE]

> > It's recommended to use the deterministic functions available in the Rollups Contracts.

Reviewer: ```suggestion
> It's recommended to use the deterministic deployment functions available in the Rollups Contracts.
```

> [!IMPORTANT]
> The commands below assumes that the following environment variables have been previous defined:
>

> > - `CONTRACT_OWNER_ADDRESS`: address of the wallet used by the application owner

Reviewer: This environment variable is not necessary.

Author: Removed at 1fbb3de

Author: All the other suggestions were accepted as well.

> - `SALT`: salt to be used by the deterministic deployment functions
> - `RPC_URL`: the RPC endpoint to be used
> - `MNEMONIC`: mnemonic phrase for the application owner's wallet (other wallet options may be used)
> - `HISTORY_FACTORY_ADDRESS`: address of a valid _History Factory_ instance
> - `AUTHORITY_ADDRESS`: address of the _Authority_ instance being used by the application

> In order to calculate the new address, the owner may call function [`calculateHistoryAddress(address, bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L35-L38).

Reviewer: It doesn't matter who calls this view function.

```suggestion
In order to calculate the address of a new history contract, one can call the view function [`calculateHistoryAddress(address,bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L35-L38).
```

This can be done with the the help of Foundry's [cast](https://book.getfoundry.sh/reference/cast/) as follows:

```shell
cast call \
    --trace --verbose \
    $HISTORY_FACTORY_ADDRESS \
    "calculateHistoryAddress(address,bytes32)(address)" \

>     $CONTRACT_OWNER_ADDRESS \

Reviewer: You want the owner of the History contract to be the Authority contract, making it the only authorized caller of the `submitClaim` function.

```suggestion
    $AUTHORITY_ADDRESS \
```

    $SALT \
    --rpc-url "$RPC_URL"
```

> Assuming the command execution is successfull, set the resulting address as a new environment variable `NEW_HISTORY_ADDRESS`, to be used later on.

Reviewer: ```suggestion
Assuming the command execution is successfull, it should print out the address of the to-be-created History contract. You may store this address in the environment variable `NEW_HISTORY_ADDRESS`, to be used later on.
```

> After that, the new instance of _History_ may be created by calling function [`newHistory(address, bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L24-L27) as follows:

Reviewer: ```suggestion
After that, the new instance of _History_ may be created by calling function [`newHistory(address,bytes32)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/history/HistoryFactory.sol#L24-L27) as follows:
```

```shell
cast send \
    --json \
    --mnemonic "$MNEMONIC" \
    $HISTORY_FACTORY_ADDRESS \
    "newHistory(address,bytes32)(History)" \

>     $CONTRACT_OWNER_ADDRESS \

Reviewer: ```suggestion
    $AUTHORITY_ADDRESS \
```

    $SALT \
    --rpc-url "$RPC_URL"
```

### 2. Replace the _History_

> [!IMPORTANT]
> The command below assumes that the environment variables set in the previous step are available, as well as  `NEW_HISTORY_ADDRESS`, which must contain the address of the new _History_.

> To replace the _History_ used by the _Authority_, call [`setHistory(_history)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/consensus/authority/Authority.sol#L65) as follows:

Reviewer: ```suggestion
To replace the _History_ used by the _Authority_, call [`setHistory(address)`](https://github.com/cartesi/rollups-contracts/blob/e8a2d82bc51167086e7928b9a6aced0d62c96cf8/onchain/rollups/contracts/consensus/authority/Authority.sol#L65) as follows:
```

```shell
cast send \
    --json \
    --mnemonic "$MNEMONIC" \
    "$AUTHORITY_ADDRESS" \
    "setHistory(address)" \
    "$NEW_HISTORY_ADDRESS" \

>     "$SALT" \

Reviewer: The function doesn't expect a salt as parameter.

```suggestion
```

> [!IMPORTANT]
> It's highly recommended that application owners upgrade their instances of the Cartesi Rollups Node to `v1.5.0`, had their applications being affected by the issue above or not.

## What has changed

The way the Cartesi Rollups Node calculates epochs has changed starting on `v1.5.0`.
This procedure is now based on the `CARTESI_EPOCH_LENGTH` environment variable instead of `CARTESI_EPOCH_DURATION`.

The value of `CARTESI_EPOCH_LENGTH` (blocks) may be derived from `CARTESI_EPOCH_DURATION` (seconds) as follows:

`CARTESI_EPOCH_LENGTH = CARTESI_EPOCH_DURATION/BLOCK_TIME`

Where `BLOCK_TIME` corresponds to the duration it takes to generate a block in the target network.

> [!TIP]
> Suppose a block is created every `12` seconds and `CARTESI_EPOCH_DURATION` is set to `86400` seconds (24 hours).
>
> So, `CARTESI_EPOCH_LENGTH = 86400/12 = 7200`

Check the [`CHANGELOG`](./CHANGELOG.md) and the [configuration](./docs/config.md) for more details.

## How to upgrade

Application owners may decide to redeploy their applications, upgrading their instances of the Cartesi Rollups Node to release `v1.5.0`.
This is the simplest way to perform the upgrade.

> In order to do so, just update the application configuration considering [what has changed](#what-has-changed) and upgrade the version of the Cartesi Rollups Node as usual.

Author: Done at 92b8b68f1862e. Please close this conversation of it's OK

> [!CAUTION]
> A redeployment will create a new instance of the application from scratch.
> All previous inputs, outputs and claims will remain associated to the previous application address, funds included.

> Alternatively, if the application owner wants to keep the application inputs, they may choose to replace the _History_ being used by the application's _Authority_ instead.

Reviewer: ```suggestion
Alternatively, the _Authority_ owner may choose to replace the _History_ being used by the application's _Authority_ instead. This path allows inputs, outputs and locked funds to remain unchanged.
```

This process is a little bit more involved and is described in the next section.

> [!NOTE]
> A new _History_ will contain no claims.

> > Once the Rollups Node is restarted with a new _History_, all previous claims will be submitted again, one-by-one, by the configured _Authority_, incurring in extra cost for the application owner.

Reviewer: It's healthy to exercise a separation between the application owner (who can change the application's consensus contract) and _Authority_ owner (who can submit claims and change the current _History_).

```suggestion
> Once the Rollups Node is restarted with a new _History_, all previous claims will be submitted again, one-by-one, by the configured _Authority_, incurring in extra cost for the _Authority_ owner.
```

## Steps to replace an Application's _History_

> [!CAUTION]
> Instances of the [History](https://github.com/cartesi/rollups-contracts/blob/v1.4.0/onchain/rollups/contracts/history/History.sol) contract from [rollups-contracts v1.4.0](https://github.com/cartesi/rollups-contracts/releases/tag/v1.4.0) may be used simultaneously by several applications through their associated _Authority_ instance.
> Application owners must consider that and exercise care when performing the steps listed below.

To keep the application inputs, before performing the upgrade to version `v1.5.0`, proceed as described in the sub-sections below.

### 1. Instantiate a new _History_

This is a two-step process.
First, the address of the new _History_ instance should be calculated.

> [!NOTE]
> It's recommended to use the deterministic deployment functions available in the Rollups Contracts.

> [!IMPORTANT]
> All commands below assume that the following environment variables have been previous defined:
>

> > - `SALT`: salt to be used by the deterministic deployment functions

Reviewer: Some users might not know what a salt is. Let's be more explicit here for those users.

```suggestion
> - `SALT`: a random 32-byte value to be used by the deterministic deployment functions
```

> - `RPC_URL`: the RPC endpoint to be used

> > - `MNEMONIC`: mnemonic phrase for the application owner's wallet (other wallet options may be used)

Reviewer: You need the mnemonic of the _Authority_ owner that to call the `setHistory` function on the _Authority_ contract. For deploying the new _History_ contract, however, any account can be used. So, might be good to rename this as `AUTHORITY_OWNER_MNEMONIC`, and have `MNEMONIC` reference any account with enough funds.

```suggestion
> - `MNEMONIC`: mnemonic phrase for the _Authority_ owner's wallet (other wallet options may be used)
```

> > - `HISTORY_FACTORY_ADDRESS`: address of a valid _History Factory_ instance

Reviewer: Would be nice to state here that for all supported networks (Ethereum, Optimism, Arbitrum, Base + respective Sepolia-based tesnets), the `HistoryFactory` is deployed at `0x1f158b5320BBf677FdA89F9a438df99BbE560A26`.

Author: Done at 92b8b68f1862e. Please close this conversation of it's OK

## What has changed

The way the Cartesi Rollups Node closes epochs has changed starting on `v1.5.0`.
This procedure is now based on the `CARTESI_EPOCH_LENGTH` environment variable instead of `CARTESI_EPOCH_DURATION`.

The value of `CARTESI_EPOCH_LENGTH` (blocks) may be derived from `CARTESI_EPOCH_DURATION` (seconds) as follows:

`CARTESI_EPOCH_LENGTH = CARTESI_EPOCH_DURATION/BLOCK_TIME`

Where `BLOCK_TIME` corresponds to the duration it takes to generate a block in the target network.

> [!TIP]
> Suppose a block is created every `12` seconds and `CARTESI_EPOCH_DURATION` is set to `86400` seconds (24 hours).
>
> So, `CARTESI_EPOCH_LENGTH = 86400/12 = 7200`

Check the [`CHANGELOG`](./CHANGELOG.md) and the [configuration](./docs/config.md) for more details.

## How to upgrade

Application owners may decide to redeploy all necessary contracts and upgrade their instances of the Cartesi Rollups Node to release `v1.5.0`.
This is the simplest way to perform the upgrade.

In order to do so, just update the application configuration considering [what has changed](#what-has-changed) and upgrade the version of the Cartesi Rollups Node as usual.

> For more details about hot to deploy an application, please refer to the [deployment process](https://docs.cartesi.io/cartesi-rollups/1.3/deployment/introduction/#deployment-process) options available in the Cartesi documentation.

Reviewer: ```suggestion
For more details about how to deploy an application, please refer to the [deployment process](https://docs.cartesi.io/cartesi-rollups/1.3/deployment/introduction/#deployment-process) options available in the Cartesi documentation.
```
